File: 56.merge-intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#

# Time:  O(nlogn)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        intervals.sort()
        for interval in intervals:
            if not res or res[-1][1] < interval[0]:
                res.append(interval)
            else:
                res[-1][1] = max(res[-1][1], interval[1])
        return res


# @lc code=end

File: 56.merge-intervals.js
/*
 * @lc app=leetcode id=56 lang=javascript
 *
 * [56] Merge Intervals
 */

// @lc code=start
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  const res = [];
  intervals = intervals.sort((a, b) =>
    a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]
  );
  for (const itvl of intervals) {
    if (!res.length || res[res.length - 1][1] < itvl[0]) {
      res.push(itvl);
    } else {
      res[res.length - 1][1] = Math.max(res[res.length - 1][1], itvl[1]);
    }
  }
  return res;
};
// @lc code=end
